python/src/iceberg/catalog/base.py

```diff
@@ -0,0 +1,223 @@
+        *,
+        namespace: Tuple[str, ...],
+        name: str,
+        schema: Schema,
+        location: Optional[str] = None,
+        partition_spec: Optional[PartitionSpec] = None,
+        properties: Optional[Dict[str, str]] = None,
+    ) -> Table:
+        """Starts a transaction and replaces the table with the provided spec.
+
+        Args:
+            namespace: A tuple of table's namespace levels. Ex: ('com','org','dept')
+            name: Table name
+            schema: Table's schema.
+            location: Location for the table. Optional Argument.
+            partition_spec: PartitionSpec for the table. Optional Argument.
+            properties: Table metadata that can be a string based dictionary. Optional Argument.
+
+        Returns:
+            Table: the replaced table instance with the updated state
+
+        Raises:
+            TableNotFoundError: If a table with the name does not exist
+        """
+
+    @abstractmethod
+    def create_namespace(self, namespace: Tuple[str, ...], properties: Optional[Dict[str, str]] = None) -> None:
+        """Create a namespace in the catalog.
+
+        Args:
+            namespace: The namespace to be created. Tuple of namespace levels. Ex: ('com','org','dept')
+            properties: A string dictionary of properties for the given namespace
+
+        Raises:
+            AlreadyExistsError: If a namespace with the name already exists in the namespace
+        """
+
+    @abstractmethod
+    def drop_namespace(self, namespace: Tuple[str, ...]) -> None:
+        """Drop a namespace.
+
+        Args:
+            namespace: The namespace to be dropped. Tuple of namespace levels. Ex: ('com','org','dept')
+
+        Raises:
+            NamespaceNotFoundError: If a namespace with the name does not exist in the namespace
+            NamespaceNotEmptyError: If the namespace is not empty
+        """
+
+    @abstractmethod
+    def list_tables(self, namespace: Optional[Tuple[str, ...]] = None) -> List[Tuple[Tuple[str, ...], str]]:
+        """List tables under the given namespace in the catalog.
+
+        If namespace not provided, will list all tables in the catalog.
+
+        Args:
+            namespace: the namespace to search. Tuple of namespace levels. Ex: ('com','org','dept')
+
+        Returns:
+            List[Tuple[str, str]]: list of tuple of table namespace and their names.
+
+        Raises:
+            NamespaceNotFoundError: If a namespace with the name does not exist in the namespace
+        """
+
+    @abstractmethod
+    def list_namespaces(self) -> List[Tuple[str, ...]]:
+        """List namespaces from the given namespace. If not given, list top-level namespaces from the catalog.
+
+        Returns:
+            List[Tuple[str, ...]]: a List of namespace, where each element is a Tuple of namespace levels. Ex: ('com','org','dept')
+        """
+
+    @abstractmethod
+    def load_namespace_metadata(self, namespace: Tuple[str, ...]) -> Dict[str, str]:
```

Comment:
How about renaming this to `load_namespace` instead? Do we need the `_metadata` part?

Reply:
IMO the model should reflect that "A catalog has namespaces that has tables". So `Catalog` --> `Namespace` --> `Table`

I thought about this several times when modeling it but didn't have the context on why it wasn't done this way in the Java API. The fact that we don't allow namespaces to be dropped that are not empty is a code smell that it should be modeled that way. What do you think?

Reply:
API is now renamed to `load_namespace` in a recent commit.